bitnami/etcd/values.yaml

```diff
@@ -23,271 +23,233 @@ image:
+## etcd configuration
+## Specify content for etcd.conf.yml
+## e.g:
+## configuration: |-
+##    foo: bar
+##    baz:
 ##
-etcd:
-  ## Initial cluster state. Allowed values: 'new' or 'existing'
-  ## If this values is not set, the default values below are set:
-  ## - 'new': when installing the chart ('helm install ...')
-  ## - 'existing': when upgrading the chart ('helm upgrade ...')
-  ##
-  initialClusterState: ""
+# configuration:
 
-## Service parameters
+## Existing ConfigMap with etcd configuration
+## NOTE: When it's set the configuration parameter is ignored
 ##
-service:
-  ## K8s service type
-  ##
-  type: ClusterIP
-  ## etcd client port
-  ##
-  port: 2379
-  ## etcd client port name override
-  ##
-  clientPortNameOverride: ""
-  ## etcd peer port
-  ##
-  peerPort: 2380
-  ## etcd peer port name override
-  ##
-  peerPortNameOverride: ""
-  ## Specify the nodePort(s) value(s) for the LoadBalancer and NodePort service types.
-  ## ref: https://kubernetes.io/docs/concepts/services-networking/service/#type-nodeport
-  ##
-  nodePorts:
-    clientPort: ""
-    peerPort: ""
-  ## Set the LoadBalancer service type to internal only.
-  ## ref: https://kubernetes.io/docs/concepts/services-networking/service/#internal-load-balancer
-  ##
-  # loadBalancerIP:
-  # loadBalancerSourceRanges: ["10.0.0.0/8"]
-  externalIPs: []
-  ## Provide any additional annotations which may be required. This can be used to
-  ## set the LoadBalancer service type to internal only.
-  ## ref: https://kubernetes.io/docs/concepts/services-networking/service/#internal-load-balancer
-  ##
-  annotations: {}
+# existingConfigmap:
 
-## Enable persistence using Persistent Volume Claims
-## ref: http://kubernetes.io/docs/user-guide/persistent-volumes/
+## An array to add extra env vars
+## e.g:
+## extraEnvVars:
+##   - name: FOO
+##     value: "bar"
 ##
-persistence:
-  ## If true, use a Persistent Volume Claim, If false, use emptyDir
+extraEnvVars: []
+
+## ConfigMap with extra environment variables
+##
+extraEnvVarsCM: ""
+
+## Secret with extra environment variables
+##
+extraEnvVarsSecret: ""
+
+## Command and args for running the container (set to default if not set). Use array form
+##
+command:
+  - /scripts/setup.sh
+args: []
+
+## Number of replicas
+##
+replicaCount: 1
+
+## Update strategy
+## ref: https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#update-strategies
+##
+updateStrategy:
+  ## Update strategy type, can be set to RollingUpdate or OnDelete.
   ##
+  type: RollingUpdate
+
+## Pod management policy
+## ref: https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#pod-management-policies
+##
+podManagementPolicy: Parallel
+
+## etcd pod host aliases
+## ref: https://kubernetes.io/docs/concepts/services-networking/add-entries-to-pod-etc-hosts-with-host-aliases/
+##
+hostAliases: []
+
+## lifecycleHooks for the etcd container to automate configuration before or after startup.
+##
+lifecycleHooks:
```

Comment:
This cannot be deactivated if you set it in the values.yaml

```
❯ cat /tmp/valuetroll.yaml
lifecycleHooks: {}

❯ helm template etcd bitnami/etcd --values /tmp/valuetroll.yaml

```
        - name: etcd
          image: docker.io/bitnami/etcd:3.4.14-debian-10-r84
          imagePullPolicy: "IfNotPresent"
          lifecycle:
            preStop:
              exec:
                command:
                - /scripts/prestop-hook.sh
```

If you have a default, then lifecycleHooks should only be for overriding.

Reply:
I was able setting `lifecycleHooks` to `nil`:
```console
$ helm template etcd bitnami/etcd --set lifecycleHooks= -s templates/statefulset.yaml
```

Reply:
More examples:

```
❯ cat /tmp/valuetroll.yaml
lifecycleHooks:
  preStop:
    http:
      - test

❯ helm template etcd . --values /tmp/valuetroll.yaml -s templates/statefulset.yaml  | grep -C 7 hook
        - name: etcd
          image: docker.io/bitnami/etcd:3.4.14-debian-10-r84
          imagePullPolicy: "IfNotPresent"
          lifecycle:
            preStop:
              exec:
                command:
                - /scripts/prestop-hook.sh
              http:
              - test

```

It's creating incompatible YAML. This would work

```
lifecycleHooks:
  preStop:
    exec:
    http:
      - test

helm template etcd . --values /tmp/valuetroll.yaml -s templates/statefulset.yaml  | grep -C 7 life
          lifecycle:
            preStop:
              http:
              - test
```

But this wouldn't

```
❯ cat /tmp/valuetroll.yaml
lifecycleHooks:
  preStop:
    exec: {}
    http:
      - test

helm template etcd . --values /tmp/valuetroll.yaml -s templates/statefulset.yaml

           lifecycle:
            preStop:
              exec:
                command:
                - /scripts/prestop-hook.sh
              http:
              - test
```

This also happens at different levels. Suppose that I don't want to have a preStop but a preStart

```
❯ cat /tmp/valuetroll.yaml
lifecycleHooks:
  preStart:
    http:
      - test
```

```
          lifecycle:
            preStart:
              http:
              - test
            preStop:
              exec:
                command:
                - /scripts/prestop-hook.sh
```

This is not something would find obvious when looking at this 

```
          lifecycle: {{- include "common.tplvalues.render" (dict "value" .Values.lifecycleHooks "context" $) | nindent 12 }}

```

Having to explicitly nullify certain sections (only by setting it like `value:`, as this `value: {}` would still render) in order to avoid incompatible YAML seems counter-intuitive to me and prone to errors. Because of this, I prefer that lifecycleHooks overrides the defaults set in the YAML itself.

Reply:
So, what do you suggest?